server/src/test/java/org/elasticsearch/cluster/coordination/Zen1IT.java

```diff
@@ -45,4 +66,202 @@ public void testZen1NodesJoiningZen2Cluster() {
+            public Settings onNodeStopped(String nodeName) {
+                String viaNode = randomValueOtherThan(nodeName, () -> randomFrom(nodes));
+                final ClusterHealthRequestBuilder clusterHealthRequestBuilder = client(viaNode).admin().cluster().prepareHealth()
+                    .setWaitForEvents(Priority.LANGUID)
+                    .setWaitForNodes(Integer.toString(nodeCount - 1))
+                    .setTimeout(TimeValue.timeValueSeconds(30));
+                if (nodeCount == 2) {
+                    clusterHealthRequestBuilder.setWaitForYellowStatus();
+                } else {
+                    clusterHealthRequestBuilder.setWaitForGreenStatus();
+                }
+                ClusterHealthResponse clusterHealthResponse = clusterHealthRequestBuilder.get();
+                assertFalse(nodeName, clusterHealthResponse.isTimedOut());
+                return Coordinator.addZen1Attribute(false, Settings.builder().put(ZEN2_SETTINGS)).build();
+            }
+        });
+
+        ensureStableCluster(nodeCount);
+        ensureGreen("test");
+        assertThat(internalCluster().size(), equalTo(nodeCount));
+    }
+
+    public void testUpgradingFromZen1ToZen2ClusterWithTwoNodes() throws Exception {
+        testRollingUpgradeFromZen1ToZen2(2);
+    }
+
+    public void testUpgradingFromZen1ToZen2ClusterWithThreeNodes() throws Exception {
+        testRollingUpgradeFromZen1ToZen2(3);
+    }
+
+    public void testUpgradingFromZen1ToZen2ClusterWithFourNodes() throws Exception {
+        testRollingUpgradeFromZen1ToZen2(4);
+    }
+
+    public void testUpgradingFromZen1ToZen2ClusterWithFiveNodes() throws Exception {
+        testRollingUpgradeFromZen1ToZen2(5);
+    }
+
+    private void testMultipleNodeMigrationFromZen1ToZen2(int nodeCount) throws Exception {
+        final List<String> oldNodes = internalCluster().startNodes(nodeCount, ZEN1_SETTINGS);
+        createIndex("test",
+            Settings.builder()
+                .put(UnassignedInfo.INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.getKey(), TimeValue.ZERO) // assign shards
+                .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, nodeCount) // causes rebalancing
+                .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, nodeCount > 1 ? 1 : 0)
+                .build());
+        ensureGreen("test");
+
+        internalCluster().startNodes(nodeCount, ZEN2_SETTINGS);
+
+        logger.info("--> updating settings to exclude old nodes");
+        client().admin().cluster().prepareUpdateSettings().setPersistentSettings(Settings.builder()
+            .put(CLUSTER_ROUTING_EXCLUDE_GROUP_SETTING.getConcreteSettingForNamespace("_name").getKey(), String.join(",", oldNodes))).get();
+
+        logger.info("--> waiting for old nodes to be vacated");
+        waitForRelocation();
+
+        while (internalCluster().size() > nodeCount) {
+            internalCluster().stopRandomNode(settings -> oldNodes.contains(NODE_NAME_SETTING.get(settings)));
```

Comment:
do we need to shut these down 1 by 1?

Reply:
Is there another way?

Reply:
unfortunately not. We can in the future add something like that to InternalTestCluster.